File: Demos/ShapePainterVB/Shapes.vb
Imports System
Imports System.Drawing
Imports System.Drawing.Drawing2D
Namespace ShapePainter

	Public Enum ShapeType
		Rectangle = 0
		Ellipse = 1
		Triangle = 2
		Quadrilateral = 3
	End Enum 

	Public MustInherit Class Shape
		Protected m_Color As Color = Color.Empty

		Public MustOverride Sub draw(ByVal g As Graphics)

		Public MustOverride ReadOnly Property Type() As ShapeType
	End Class 

	Public Class RectangleShape
	Inherits Shape
		Private m_Rectangle As Rectangle

		Public Sub New(ByVal x As Integer, ByVal y As Integer, ByVal width As Integer, ByVal height As Integer, ByVal color As Color)
			Me.m_Rectangle = New Rectangle(x, y, width, height)
			Me.m_Color = color
		End Sub

		Public Overloads Overrides ReadOnly Property Type() As ShapeType
			Get
				Return ShapeType.Rectangle
			End Get
		End Property

		Public Shared Function CreateFromRandom(ByVal random As Random, ByVal size As Size, ByVal drawingBoardSize As Size) As RectangleShape
            Dim width As Integer = random.Next(10, size.Width)
			Dim height As Integer = random.Next(10, size.Height)
            Dim x As Integer = random.Next(Math.Max(0, drawingBoardSize.Width - width))
            Dim y As Integer = random.Next(Math.Max(0, drawingBoardSize.Height - height))
            Dim color As Color = Drawing.Color.FromArgb(0, 0, random.Next(128, 255))
			Dim rectangle As RectangleShape = New RectangleShape(x, y, width - 5, height - 5, color)
			Return rectangle
		End Function

		Public Overloads Overrides Sub draw(ByVal g As Graphics)
			g.FillRectangle(New SolidBrush(Me.m_Color), Me.m_Rectangle)
		End Sub
	End Class 

	Public Class EllipseShape
	Inherits Shape
		Private m_Ellipse As Rectangle

		Public Sub New(ByVal x As Integer, ByVal y As Integer, ByVal width As Integer, ByVal height As Integer, ByVal color As Color)
			Me.m_Ellipse = New Rectangle(x, y, width, height)
			Me.m_Color = color
		End Sub

		Public Overloads Overrides ReadOnly Property Type() As ShapeType
			Get
				Return ShapeType.Ellipse
			End Get
		End Property

		Public Shared Function CreateFromRandom(ByVal random As Random, ByVal size As Size, ByVal drawingBoardSize As Size) As EllipseShape
			Dim width As Integer = random.Next(10, size.Width)
			Dim height As Integer = random.Next(10, size.Height)
            Dim x As Integer = random.Next(Math.Max(0, drawingBoardSize.Width - width))
            Dim y As Integer = random.Next(Math.Max(0, drawingBoardSize.Height - height))
            Dim color As Color = Drawing.Color.FromArgb(0, random.Next(128, 255), 0)
			Return New EllipseShape(x, y, width - 5, height - 5, color)
		End Function

		Public Overloads Overrides Sub draw(ByVal g As Graphics)
			g.SmoothingMode = SmoothingMode.AntiAlias
			g.FillEllipse(New SolidBrush(Me.m_Color), Me.m_Ellipse)
		End Sub
	End Class 

	Public Class TriangleShape
	Inherits Shape
		Private m_Triangle As Point() = New Point(3) {}

		Public Sub New(ByVal triangle As Point(), ByVal color As Color)
			Me.m_Triangle = triangle
			Me.m_Color = color
		End Sub

		Public Overloads Overrides ReadOnly Property Type() As ShapeType
			Get
				Return ShapeType.Triangle
			End Get
		End Property

		Public Shared Function CreateFromRandom(ByVal random As Random, ByVal size As Size, ByVal drawingBoardSize As Size) As TriangleShape
            Dim triangle(2) As Point
            Dim x_offset As Integer = random.Next(Math.Max(0, drawingBoardSize.Width - size.Width))
            Dim y_offset As Integer = random.Next(Math.Max(0, drawingBoardSize.Height - size.Height))
            Dim i As Integer
            For i = 0 To 2
                triangle(i) = New Point(random.Next(1, size.Width) + x_offset, random.Next(1, size.Height) + y_offset)
            Next
            Dim color As Color = Drawing.Color.FromArgb(random.Next(128, 255), 0, 0)
            Return New TriangleShape(triangle, color)
		End Function

		Public Overloads Overrides Sub draw(ByVal g As Graphics)
			g.SmoothingMode = SmoothingMode.AntiAlias
			g.FillPolygon(New SolidBrush(Me.m_Color), Me.m_Triangle)
		End Sub
	End Class 
End Namespace

File: Demos/ShapePainter/Shapes.cs
using System;
using System.Drawing;
using System.Drawing.Drawing2D;

namespace ShapePainter
{
	public enum ShapeType
	{
		Rectangle,
		Ellipse,
		Triangle,
		Quadrilateral
	}
	/// <summary>
	/// Summary description for Shape.
	/// </summary>
	public abstract class Shape
	{
		protected Color m_Color = Color.Empty;
		public abstract void draw(Graphics g);
		public abstract ShapeType Type
		{
			get;
		}
	}

	public class RectangleShape : Shape
	{
		private Rectangle m_Rectangle;

		public RectangleShape(int x, int y, int width, int height, Color color)
		{
			this.m_Rectangle = new Rectangle(x, y, width, height);
			this.m_Color = color;
		}

		public override ShapeType Type
		{
			get
			{
				return ShapeType.Rectangle;
			}
		}

		public static RectangleShape CreateFromRandom(Random random, Size size, Size drawingBoardSize)
		{
			int width = random.Next(10, size.Width);
			int height = random.Next(10, size.Height);
			int x = random.Next(Math.Max(0, drawingBoardSize.Width - width));
			int y = random.Next(Math.Max(0, drawingBoardSize.Height - height));
			Color color = Color.FromArgb(0, 0, random.Next(128, 255));
			RectangleShape rectangle = new RectangleShape(x, y, width - 5, height - 5, color);
			return rectangle;
		}
		
		public override void draw(Graphics g)
		{
			g.FillRectangle(new SolidBrush(this.m_Color), this.m_Rectangle);
		}
	}

	public class EllipseShape : Shape
	{
		private Rectangle m_Ellipse;

		public EllipseShape(int x, int y, int width, int height, Color color)
		{
			this.m_Ellipse = new Rectangle(x, y, width, height);
			this.m_Color = color;
		}

		public override ShapeType Type
		{
			get
			{
				return ShapeType.Ellipse;
			}
		}

		public static EllipseShape CreateFromRandom(Random random, Size size, Size drawingBoardSize)
		{
			int width = random.Next(10, size.Width);
			int height = random.Next(10, size.Height);
			int x = random.Next(Math.Max(0, drawingBoardSize.Width - width));
			int y = random.Next(Math.Max(0, drawingBoardSize.Height - height));
			Color color = Color.FromArgb(0, random.Next(128, 255), 0);
			return new EllipseShape(x, y, width - 5, height - 5, color);
		}

		public override void draw(Graphics g)
		{
			g.SmoothingMode = SmoothingMode.AntiAlias;
			g.FillEllipse(new SolidBrush(this.m_Color), this.m_Ellipse);
		}
	}

	public class TriangleShape : Shape
	{
		Point[] m_Triangle = new Point[3];

		public TriangleShape(Point[] triangle, Color color)
		{
			this.m_Triangle = triangle;
			this.m_Color = color;
		}

		public override ShapeType Type
		{
			get
			{
				return ShapeType.Triangle;
			}
		}

		public static TriangleShape CreateFromRandom(Random random, Size size, Size drawingBoardSize)
		{
			Point[] triangle = new Point[3];
			int x_offset = random.Next(Math.Max(0, drawingBoardSize.Width - size.Width));
			int y_offset = random.Next(Math.Max(0, drawingBoardSize.Height - size.Height));
			for (int i = 0; i < 3; i++)
			{
				triangle[i] = new Point(random.Next(1, size.Width) + x_offset, random.Next(1, size.Height) + y_offset);
			}

			Color color = Color.FromArgb(random.Next(128, 255), 0, 0);
			return new TriangleShape(triangle, color);
		}

		public override void draw(Graphics g)
		{
			g.SmoothingMode = SmoothingMode.AntiAlias;
			g.FillPolygon(new SolidBrush(this.m_Color), this.m_Triangle);
		}
	}
}
